zcash_client_backend/src/data_api/chain.rs

```diff
@@ -149,61 +180,81 @@ where
             .map(|(tag, ivk)| (tag, PreparedIncomingViewingKey::new(&ivk))),
     );
 
-    // Start at either the provided height, or where we synced up to previously.
-    let (scan_from, mut prior_block_metadata) = match from_height {
-        Some(h) => {
-            // if we are provided with a starting height, obtain the metadata for the previous
-            // block (if any is available)
-            (
-                Some(h),
-                if h > BlockHeight::from(0) {
-                    data_db.block_metadata(h - 1).map_err(Error::Wallet)?
-                } else {
-                    None
-                },
-            )
-        }
-        None => {
-            let last_scanned = data_db.block_fully_scanned().map_err(Error::Wallet)?;
-            last_scanned.map_or_else(|| (None, None), |m| (Some(m.block_height + 1), Some(m)))
-        }
+    let mut prior_block_metadata = if from_height > BlockHeight::from(0) {
+        data_db
+            .block_metadata(from_height - 1)
+            .map_err(Error::Wallet)?
+    } else {
+        None
     };
 
-    block_source.with_blocks::<_, DbT::Error>(scan_from, limit, |block: CompactBlock| {
-        add_block_to_runner(params, block, &mut batch_runner);
-        Ok(())
-    })?;
+    let mut continuity_check_metadata = prior_block_metadata;
+    block_source.with_blocks::<_, DbT::Error>(
+        Some(from_height),
+        Some(limit),
+        |block: CompactBlock| {
+            // check block continuity
+            if let Some(scan_error) = check_continuity(&block, continuity_check_metadata.as_ref()) {
+                return Err(Error::Scan(scan_error));
+            }
+            continuity_check_metadata = continuity_check_metadata.as_ref().map(|m| {
```

Comment:
There's a small issue here: if the user calls this method with `from_height = 0`, `prior_block_metadata` was `None`, which means that `continuity_check_metadata` is initialized as `None`, and this `map` propagates that for all time, even if every block in `block_source` has `sapling_commitment_tree_size` set correctly.

The fix here is to recognise that if `from_height = 0`, then the previous block's continuity metadata implicitly has `sapling_commitment_tree_size = 0`, so we can always derive a correct `continuity_check_metadata` for height 0 itself. So we should special-case this height for correctness (of downstream usage on chains that activated Sapling from genesis):
```suggestion
            if from_height == BlockHeight::from(0) {
                // We can always derive a valid `continuity_check_metadata` for the
                // genesis block, even if the block source doesn't have
                // `sapling_commitment_tree_size`. So briefly set it to a dummy value that
                // ensures the `map` below produces the correct genesis block value.
                assert!(continuity_check_metadata.is_none());
                continuity_check_metadata =
                    Some(BlockMetadata::from_parts(0, BlockHash([0; 32]), 0));
            }
            continuity_check_metadata = continuity_check_metadata.as_ref().map(|m| {
```